pkg/volume/flexvolume/probe_test.go

```diff
@@ -0,0 +1,294 @@
+
+	// Assert
+	assert.True(t, updated)
+	assert.NoError(t, err)
+	for i := 0; i < 4; i++ {
+		updated, _, err = prober.Probe()
+		assert.False(t, updated)
+		assert.NoError(t, err)
+	}
+}
+
+// When many events are triggered quickly in succession, events should stop triggering a probe update
+// after a certain limit.
+func TestProberRateLimit(t *testing.T) {
+	// Arrange
+	driverPath, _, watcher, prober := initTestEnvironment(t)
+	for i := 0; i < watchEventLimit; i++ {
+		watcher.TriggerEvent(fsnotify.Write, path.Join(driverPath, driverName))
+	}
+
+	// Act
+	updated, plugins, err := prober.Probe()
+
+	// Assert
+	assert.True(t, updated)
+	assert.Equal(t, 1, len(plugins))
+	assert.NoError(t, err)
+
+	// Arrange
+	watcher.TriggerEvent(fsnotify.Write, path.Join(driverPath, driverName))
+
+	// Act
+	updated, _, err = prober.Probe()
+
+	// Assert
+	assert.False(t, updated)
+	assert.NoError(t, err)
+}
+
+func TestProberError(t *testing.T) {
+	fs := utilfs.NewFakeFs()
+	watcher := NewMockWatcher()
+	prober := &flexVolumeProber{
+		pluginDir: pluginDir,
+		watcher:   watcher,
+		fs:        fs,
+		factory:   mockPluginFactory{error: true},
+	}
+	installDriver(driverName, fs)
+	prober.Init()
+
+	_, _, err := prober.Probe()
+	assert.Error(t, err)
+}
+
+// Installs a mock driver (an empty file) in the mock fs.
+func installDriver(driverName string, fs utilfs.Filesystem) {
+	driverPath := path.Join(pluginDir, driverName)
+	fs.MkdirAll(driverPath, 0666)
+	fs.Create(path.Join(driverPath, driverName))
+}
+
+// Initializes mocks, installs a single driver in the mock fs, then initializes prober.
+func initTestEnvironment(t *testing.T) (
+	driverPath string,
+	fs utilfs.Filesystem,
+	watcher *mockWatcher,
+	prober volume.DynamicPluginProber) {
+	fs = utilfs.NewFakeFs()
+	watcher = NewMockWatcher()
+	prober = &flexVolumeProber{
+		pluginDir: pluginDir,
+		watcher:   watcher,
+		fs:        fs,
+		factory:   mockPluginFactory{error: false},
+	}
+	driverPath = path.Join(pluginDir, driverName)
+	installDriver(driverName, fs)
+	prober.Init()
+
+	assert.NotNilf(t, watcher.eventHandler,
+		"Expect watch event handler to be registered after prober init, but is not.")
+	return
+}
+
+// Mock filesystem watcher
+type mockWatcher struct {
+	watches      []string // List of watches added by the prober, ordered from least recent to most recent.
+	eventHandler utilfs.FSEventHandler
+}
+
+var _ utilfs.FSWatcher = &mockWatcher{}
+
+func NewMockWatcher() *mockWatcher {
+	return &mockWatcher{
+		watches: nil,
+	}
+}
+
+func (w *mockWatcher) Init(eventHandler utilfs.FSEventHandler, _ utilfs.FSErrorHandler) {
+	w.eventHandler = eventHandler
+}
+
+func (w *mockWatcher) AddWatch(path string) error {
+	w.watches = append(w.watches, path)
+	return nil
+}
+
+// Triggers a mock filesystem event.
+func (w *mockWatcher) TriggerEvent(op fsnotify.Op, filename string) {
+	w.eventHandler(fsnotify.Event{Op: op, Name: filename})
+}
+
+// Mock Flexvolume plugin
+type mockPluginFactory struct {
```

Comment:
Rename this to fakePluginFactory and move to a separate file. We typically put the fakes in the same package as the real ones, but in a different file.

Reply:
ACK